### PR TITLE
[YAML] - Kafka Read and Write proto

### PR DESCRIPTION
**Pull request by ffernandez92**
- Include the option to read and write proto in YAML Beam [KAFKA]

addresses #28664

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by ffernandez92** [comment]
Run PythonDocker PreCommit 3.9

**Comment by ffernandez92** [comment]
Run PythonDocker PreCommit 3.8

**Comment by ffernandez92** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by ffernandez92** [comment]
Run PythonDocker PreCommit 3.8

**Comment by ffernandez92** [comment]
I believe some other Go PR broke the Python Docker tests. @brucearctor , this PR enables Kafka Beam YAML to read and write proto.

Reading is 'easy' because the user has to provide the file descriptor and the message name so we can build the Row from the bytes.

Writing was a bit more challenging. I initially had this method to write to proto by just providing the Row schema:

```
public static SerializableFunction<Row, byte[]> getRowToProtoBytes() {
    return new SimpleFunction<Row, byte[]>() {
      @Override
      public byte[] apply(Row input) {
        SchemaApi.Row rowProto = SchemaTranslation.rowToProto(input);
        return rowProto.toByteArray();
      }
    };
  }
```
However, this generates a SchemaApi.Row proto. This implies that if another system intends to read these protos, one must construct the proto using the SchemaApi file descriptor, etc. This approach doesn't make much sense, as we don't want users to remember this as the output.

As a result, I am also requesting the file descriptor and message name for the output. This enables other systems to access the output proto without needing to remember intricate details. While this description may be dense, I believe it makes sense. I tested it with Dataflow (both read and write), and it worked fine.

File: sdks/java/io/kafka/src/test/resources/proto_byte/file_descriptor/proto_byte_utils.pb
Comment (brucearctor) [inline sdks/java/io/kafka/src/test/resources/proto_byte/file_descriptor/proto_byte_utils.pb:1]: Do we want to commit this file as is -- or generate it based on `sdks/java/io/kafka/src/test/resources/proto_byte/proto_byte_utils.proto` as part of the CI/testing process?  Probably fine as is for now, but hoping we consider that as an option.

**Comment by brucearctor** [comment]
@alexvanboxel -- you spent LOTs of time thinking about beam/proto.  Can you comment on the workflow/usability?  Thanks!

File: sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoByteUtils.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.Row;
+import org.apache.commons.compress.utils.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for working with Protocol Buffer (Proto) data in the context of Apache Beam. This
```
Comment (brucearctor) [inline sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoByteUtils.java:45]: nit:  is 'in the context of Apache Beam' unneeded, given where the code exists?

**Comment by brucearctor** [comment]
> I believe some other Go PR broke the Python Docker tests. @brucearctor , this PR enables Kafka Beam YAML to read and write proto.
> 
> Reading is 'easy' because the user has to provide the file descriptor and the message name so we can build the Row from the bytes.
> 
> Writing was a bit more challenging. I initially had this method to write to proto by just providing the Row schema:
> 
> ```
> public static SerializableFunction<Row, byte[]> getRowToProtoBytes() {
>     return new SimpleFunction<Row, byte[]>() {
>       @Override
>       public byte[] apply(Row input) {
>         SchemaApi.Row rowProto = SchemaTranslation.rowToProto(input);
>         return rowProto.toByteArray();
>       }
>     };
>   }
> ```
> 
> However, this generates a SchemaApi.Row proto. This implies that if another system intends to read these protos, one must construct the proto using the SchemaApi file descriptor, etc. This approach doesn't make much sense, as we don't want users to remember this as the output.
> 
> As a result, I am also requesting the file descriptor and message name for the output. This enables other systems to access the output proto without needing to remember intricate details. While this description may be dense, I believe it makes sense. I tested it with Dataflow (both read and write), and it worked fine.

Finally finding a few moments to look again a bit closer.  

(a) As mentioned, the general approach makes sense to me.  

(b) i imagine you are correct about failing tests, though haven't dug into that yet.

I don't know the extent that I highlighted to you previously, but Scio had a creative way for handling proto bytes ...  Do checkout: https://spotify.github.io/scio/io/Protobuf.html and related, just for additional context on another way could be handled.  I think there are pros/cons.

**Comment by ffernandez92** [comment]
Run PythonDocker PreCommit 3.8

**Comment by ffernandez92** [comment]
I tried reading and writing in proto format with DF and it works fine. Besides that I ran into an issue when trying to write into BQ but that's unrelated to proto. These lines produce a null pointer 

```
        Integer numStreams = configuration.getNumStreams();
        // Triggering frequency is only applicable for exactly-once
        if (!configuration.getUseAtLeastOnceSemantics()) {
          write =
              write.withTriggeringFrequency(
                  (triggeringFrequency == null || triggeringFrequency <= 0)
                      ? DEFAULT_TRIGGERING_FREQUENCY
                      : Duration.standardSeconds(triggeringFrequency));
        }
        // set num streams if specified, otherwise default to autoSharding
        if (numStreams > 0) {
          write = write.withNumStorageWriteApiStreams(numStreams);
        } else if (autoSharding == null || autoSharding) {
          write = write.withAutoSharding();
        }
``` 
numStreams and configuration.getUseAtLeastOnceSemantics don't have default values and are not mapped to standard_io.yml 

@robertwb should I open an issue to fix those and make them fallback to a default value or is someone else working on that? 

Screenshot 2023-11-10 at 13 02 29

**Comment by brucearctor** [comment]
Open an issue, so it doesn't get forgotten.   It someone us addresses thwn the issue could get closed as a duplicate.  

Seems distinct from the current scope of this PR.

**Comment by brucearctor** [comment]
Run PythonDocker PreCommit 3.8

**Comment by brucearctor** [comment]
@ffernandez92 -- do we want to think about schema registry for kafka as separate from this?  

Ex:  if writing to a topic with an associated schema registry, can we get the schema from that, rather than passing in the descriptor?

**Comment by brucearctor** [comment]
the [ at least currently ] failing tests look not related to this code change, so I am comfortable with that aspect of things.

**Review by brucearctor — APPROVED** [review]
LGTM -- addresses our needs.  We can always refactor and make better.

**Pull request merged**